Error while checking build/O2/fullCI for 7acd343810ce0527d2317d2c6eeb87eeed45d25d at 2023-03-29 16:14:
```
## sw/BUILD/O2-latest/log
/sw/SOURCES/O2/11054-slc8_x86-64/0/Framework/Core/src/runDataProcessing.cxx:2638:35: error: invalid initialization of reference of type 'o2::framework::DriverInfo&' from expression of type 'o2::framework::DriverControl'
ninja: build stopped: subcommand failed.
```
Full log [here](https://ali-ci.cern.ch/alice-build-logs/AliceO2Group/AliceO2/11054/7acd343810ce0527d2317d2c6eeb87eeed45d25d/build_O2_fullCI/pretty.html).

--- a/Framework/Core/include/Framework/DriverControl.h
+++ b/Framework/Core/include/Framework/DriverControl.h
@@ -20,9 +20,7 @@
 #include "Framework/DeviceSpec.h"
 #include "Framework/DeviceExecution.h"
 
-namespace o2
-{
-namespace framework
+namespace o2::framework
 {
 
 /// These are the possible states for the driver controller
@@ -53,7 +51,6 @@ struct DriverControl {
   bool defaultStopped;
 };
 
-} // namespace framework
 } // namespace o2
 
 #endif

--- a/Framework/Core/include/Framework/DriverInfo.h
+++ b/Framework/Core/include/Framework/DriverInfo.h
@@ -164,6 +164,8 @@ struct DriverInfo {
 
   /// The last error reported by the driver itself
   std::string lastError;
+  /// Driver mode
+  DriverMode mode = DriverMode::STANDALONE;
 };
 
 struct DriverInfoHelper {

--- a/Framework/Core/include/Framework/ProcessingPolicies.h
+++ b/Framework/Core/include/Framework/ProcessingPolicies.h
@@ -34,6 +34,16 @@ struct ProcessingPolicies {
   enum EarlyForwardPolicy earlyForward;
 };
 
+/// The mode in which the driver is running. Should be MASTER when running locally,
+/// EMBEDDED when running under a control system.
+enum struct DriverMode {
+  /// Default mode. The master is responsible for spawning and controlling the devices.
+  STANDALONE,
+  /// The master is also running under a control system, so it does not control the devices, however, it does
+  /// receive metrics and information from them and displays it via the RemoteGUI.
+  EMBEDDED
+};
+
 } // namespace o2::framework
 
 #endif // O2_FRAMEWORK_PROCESSING_POLICIES_H_

--- a/Framework/Core/src/DDSConfigHelpers.cxx
+++ b/Framework/Core/src/DDSConfigHelpers.cxx
@@ -112,6 +112,7 @@ std::string xmlEncode(std::string const& source)
 }
 
 void DDSConfigHelpers::dumpDeviceSpec2DDS(std::ostream& out,
+                                          DriverMode driverMode,
                                           std::string const& workflowSuffix,
                                           std::vector<DataProcessorSpec> const& workflow,
                                           std::vector<DataProcessorInfo> const& dataProcessorInfos,
@@ -167,6 +168,19 @@ void DDSConfigHelpers::dumpDeviceSpec2DDS(std::ostream& out,
       out << "   "
           << fmt::format("<property name=\"fmqchan_{}\" />\n", rewriter.properties[rewriter.names[ci]].value);
     }
+  }
+
+  if (driverMode == DriverMode::EMBEDDED) {
+    out << "   "
+        << fmt::format("<decltask name=\"{}{}\">\n", "dplDriver", workflowSuffix);
+    out << "       "
+        << fmt::format(R"(<assets><name>dpl_json{}</name></assets>)", workflowSuffix) << "\n";
+    out << "       "
+        << R"(<exe reachable="true">)";
+    out << fmt::format("cat ${{DDS_LOCATION}}/dpl_json{}.asset | o2-dpl-run --driver-mode embedded", workflowSuffix);
+    out << R"(</exe>)"
+        << "\n";
+    out << "</decltask>";
   }
 
   for (size_t di = 0; di < specs.size(); ++di) {

--- a/Framework/Core/src/DDSConfigHelpers.h
+++ b/Framework/Core/src/DDSConfigHelpers.h
@@ -16,6 +16,7 @@
 #include "Framework/CommandInfo.h"
 #include "Framework/DataProcessorSpec.h"
 #include "Framework/DataProcessorInfo.h"
+#include "Framework/ProcessingPolicies.h"
 #include <vector>
 #include <iosfwd>
 #include <string>
@@ -26,6 +27,7 @@ namespace o2::framework
 /// Helper to dump DDS configuration to run in a deployed
 /// manner.
 /// @a out is a stream where the configuration will be printed
+/// @a mode is the mode in which the workflow is being run
 /// @a workflow is the workflow to dump
 /// @a metadata is the metadata to dump
 /// @a workflowSuffix a suffix to add to all the ids in the workflow
@@ -36,6 +38,7 @@ namespace o2::framework
 /// @a the full command being used
 struct DDSConfigHelpers {
   static void dumpDeviceSpec2DDS(std::ostream& out,
+                                 DriverMode mode,
                                  std::string const& workflowSuffix,
                                  std::vector<DataProcessorSpec> const& workflow,
                                  std::vector<DataProcessorInfo> const& metadata,

--- a/Framework/Core/src/ProcessingPoliciesHelpers.cxx
+++ b/Framework/Core/src/ProcessingPoliciesHelpers.cxx
@@ -29,6 +29,32 @@ std::istream& operator>>(std::istream& in, enum TerminationPolicy& policy)
   return in;
 }
 
+std::ostream& operator<<(std::ostream& out, const enum DriverMode& mode)
+{
+  if (mode == DriverMode::STANDALONE) {
+    out << "standalone";
+  } else if (mode == DriverMode::EMBEDDED) {
+    out << "embedded";
+  } else {
+    out.setstate(std::ios_base::failbit);
+  }
+  return out;
+}
+
+std::istream& operator>>(std::istream& in, enum DriverMode& mode)
+{
+  std::string token;
+  in >> token;
+  if (token == "standalone") {
+    mode = DriverMode::STANDALONE;
+  } else if (token == "embedded") {
+    mode = DriverMode::EMBEDDED;
+  } else {
+    in.setstate(std::ios_base::failbit);
+  }
+  return in;
+}
+
 std::ostream& operator<<(std::ostream& out, const enum TerminationPolicy& policy)
 {
   if (policy == TerminationPolicy::QUIT) {

--- a/Framework/Core/src/ProcessingPoliciesHelpers.h
+++ b/Framework/Core/src/ProcessingPoliciesHelpers.h
@@ -25,6 +25,8 @@ std::istream& operator>>(std::istream& in, enum EarlyForwardPolicy& policy);
 std::ostream& operator<<(std::ostream& out, const enum EarlyForwardPolicy& policy);
 std::istream& operator>>(std::istream& in, enum LogParsingHelpers::LogLevel& level);
 std::ostream& operator<<(std::ostream& out, const enum LogParsingHelpers::LogLevel& level);
+std::istream& operator>>(std::istream& in, enum DriverMode& level);
+std::ostream& operator<<(std::ostream& out, const enum DriverMode& level);
 } // namespace o2::framework
 
 #endif // O2_FRAMEWORK_PROCESSINGPOLICIESHELPERS_H_

--- a/Framework/Core/src/runDataProcessing.cxx
+++ b/Framework/Core/src/runDataProcessing.cxx
@@ -1735,7 +1735,7 @@ int runStateMachine(DataProcessorSpecs const& workflow,
           auto& context = childFds[di];
           createPipes(context.childstdin);
           createPipes(context.childstdout);
-          if (runningWorkflow.devices[di].resource.hostname != driverInfo.deployHostname) {
+          if (driverInfo.mode == DriverMode::EMBEDDED || runningWorkflow.devices[di].resource.hostname != driverInfo.deployHostname) {
             spawnRemoteDevice(forwardedStdin.str(),
                               runningWorkflow.devices[di], controls[di], deviceExecutions[di], infos);
           } else {
@@ -2131,6 +2131,7 @@ void overrideLabels(ConfigContext& ctx, WorkflowSpec& workflow)
 
 /// Helper function to initialise the controller from the command line options.
 void initialiseDriverControl(bpo::variables_map const& varmap,
+                             DriverInfo& driverInfo,
                              DriverControl& control)
 {
   // Control is initialised outside the main loop because
@@ -2166,16 +2167,17 @@ void initialiseDriverControl(bpo::variables_map const& varmap,
     // because DDS needs to be able to have actual Executions in
     // order to provide a correct configuration.
     control.callbacks = {[filename = varmap["dds"].as<std::string>(),
-                          workflowSuffix = varmap["dds-workflow-suffix"]](WorkflowSpec const& workflow,
-                                                                          DeviceSpecs const& specs,
-                                                                          DeviceExecutions const& executions,
-                                                                          DataProcessorInfos& dataProcessorInfos,
-                                                                          CommandInfo const& commandInfo) {
+                          workflowSuffix = varmap["dds-workflow-suffix"],
+                          driverMode = driverInfo.mode](WorkflowSpec const& workflow,
+                                                        DeviceSpecs const& specs,
+                                                        DeviceExecutions const& executions,
+                                                        DataProcessorInfos& dataProcessorInfos,
+                                                        CommandInfo const& commandInfo) {
       if (filename == "-") {
-        DDSConfigHelpers::dumpDeviceSpec2DDS(std::cout, workflowSuffix.as<std::string>(), workflow, dataProcessorInfos, specs, executions, commandInfo);
+        DDSConfigHelpers::dumpDeviceSpec2DDS(std::cout, driverMode, workflowSuffix.as<std::string>(), workflow, dataProcessorInfos, specs, executions, commandInfo);
       } else {
         std::ofstream out(filename);
-        DDSConfigHelpers::dumpDeviceSpec2DDS(out, workflowSuffix.as<std::string>(), workflow, dataProcessorInfos, specs, executions, commandInfo);
+        DDSConfigHelpers::dumpDeviceSpec2DDS(out, driverMode, workflowSuffix.as<std::string>(), workflow, dataProcessorInfos, specs, executions, commandInfo);
       }
     }};
     control.forcedTransitions = {
@@ -2358,6 +2360,7 @@ int doMain(int argc, char** argv, o2::framework::WorkflowSpec const& workflow,
   enum LogParsingHelpers::LogLevel minFailureLevel;
   bpo::options_description executorOptions("Executor options");
   const char* helpDescription = "print help: short, full, executor, or processor name";
+  enum DriverMode driverMode;
   executorOptions.add_options()                                                                                                                                        //
     ("help,h", bpo::value<std::string>()->implicit_value("short"), helpDescription)                                                                                    //                                                                                                       //
     ("quiet,q", bpo::value<bool>()->zero_tokens()->default_value(false), "quiet operation")                                                                            //                                                                                                         //
@@ -2383,6 +2386,7 @@ int doMain(int argc, char** argv, o2::framework::WorkflowSpec const& workflow,
     ("dds-workflow-suffix,D", bpo::value<std::string>()->default_value(""), "suffix for DDS names")                                                                    //                                                                                                                                  //
     ("dump-workflow,dump", bpo::value<bool>()->zero_tokens()->default_value(false), "dump workflow as JSON")                                                           //                                                                                                                                    //
     ("dump-workflow-file", bpo::value<std::string>()->default_value("-"), "file to which do the dump")                                                                 //                                                                                                                                      //
+    ("driver-mode", bpo::value<DriverMode>(&driverMode)->default_value(DriverMode::STANDALONE), R"(how to run the driver. default: "standalone". Valid: "embedded")")  //                                                                                                                                      //
     ("run", bpo::value<bool>()->zero_tokens()->default_value(false), "run workflow merged so far. It implies --batch. Use --no-batch to see the GUI")                  //                                                                                                                                        //
     ("no-IPC", bpo::value<bool>()->zero_tokens()->default_value(false), "disable IPC topology optimization")                                                           //                                                                                                                                        //
     ("o2-control,o2", bpo::value<std::string>()->default_value(""), "dump O2 Control workflow configuration under the specified name")                                 //
@@ -2631,8 +2635,6 @@ int doMain(int argc, char** argv, o2::framework::WorkflowSpec const& workflow,
       exit(1);
     }
   }
-  DriverControl driverControl;
-  initialiseDriverControl(varmap, driverControl);
 
   auto evaluateBatchOption = [&varmap]() -> bool {
     if (varmap.count("no-batch") > 0) {
@@ -2665,6 +2667,8 @@ int doMain(int argc, char** argv, o2::framework::WorkflowSpec const& workflow,
   driverInfo.noSHMCleanup = varmap["no-cleanup"].as<bool>();
   driverInfo.processingPolicies.termination = varmap["completion-policy"].as<TerminationPolicy>();
   driverInfo.processingPolicies.earlyForward = varmap["early-forward-policy"].as<EarlyForwardPolicy>();
+  driverInfo.mode = varmap["driver-mode"].as<DriverMode>();
+
   if (varmap["error-policy"].defaulted() && driverInfo.batch == false) {
     driverInfo.processingPolicies.error = TerminationPolicy::WAIT;
   } else {
@@ -2681,6 +2685,9 @@ int doMain(int argc, char** argv, o2::framework::WorkflowSpec const& workflow,
   // FIXME: should use the whole dataProcessorInfos, actually...
   driverInfo.processorInfo = dataProcessorInfos;
   driverInfo.configContext = &configContext;
+
+  DriverControl driverControl;
+  initialiseDriverControl(varmap, driverInfo, driverControl);
 
   commandInfo.merge(CommandInfo(argc, argv));
 

--- a/Framework/Core/test/test_FrameworkDataFlowToDDS.cxx
+++ b/Framework/Core/test/test_FrameworkDataFlowToDDS.cxx
@@ -151,7 +151,7 @@ TEST_CASE("TestDDS")
                                       devices, executions, controls,
                                       "workflow-id");
   CommandInfo command{"foo"};
-  DDSConfigHelpers::dumpDeviceSpec2DDS(ss, "", workflow, dataProcessorInfos, devices, executions, command);
+  DDSConfigHelpers::dumpDeviceSpec2DDS(ss, DriverMode::STANDALONE, "", workflow, dataProcessorInfos, devices, executions, command);
   auto expected = R"EXPECTED(<topology name="o2-dataflow">
 <asset name="dpl_json" type="inline" visibility="global" value="{
     &quot;workflow&quot;: [
@@ -406,7 +406,7 @@ TEST_CASE("TestDDSExpendable")
                                       devices, executions, controls,
                                       "workflow-id");
   CommandInfo command{"foo"};
-  DDSConfigHelpers::dumpDeviceSpec2DDS(ss, "", workflow, dataProcessorInfos, devices, executions, command);
+  DDSConfigHelpers::dumpDeviceSpec2DDS(ss, DriverMode::STANDALONE, "", workflow, dataProcessorInfos, devices, executions, command);
   auto expected = R"EXPECTED(<topology name="o2-dataflow">
 <declrequirement name="odc_expendable_task" type="custom" value="true" />
 <asset name="dpl_json" type="inline" visibility="global" value="{

